### PR TITLE
Changes for module 5

### PR DESCRIPTION
**Pull request by hugh-emerson**
Opening this PR for you as you originally opened one against the CorndelWithSoftwire repo by mistake.

I've targeted the module 4 branch so only the relevant changes are shown. If you merge the module 4 branch before this one, then you'll need to merge it into master.

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+.env
+.env.template
+.vscode
+.gitignore
+README.md
```
Comment (hugh-emerson) [inline .dockerignore:5]: This list is reasonable - you've included the most important file, `.env`. But some more files/folders could be added. E.g.: `.vagrant`, `Vagrantfile`, `__pycache__`, `.git`, `Dockerfile`

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+version: "3.3"
+services:
+  app:
+    image: todo-app:dev
+    ports:
+      - 5000:5000
+    env_file: 
+      - .env
```
Comment (hugh-emerson) [inline docker-compose.yml:8]: You don't actually need this here since .env will be included in the bind mount.

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+version: "3.3"
+services:
+  app:
+    image: todo-app:dev
```
Comment (hugh-emerson) [inline docker-compose.yml:4]: This currently relies on you running the correct `docker build ...` command to produce the correct `todo-app:dev` image.

You can instead do this to get docker-compose to handle the building
```suggestion
    build: 
      context: .
      target: development
```

Use `docker-compose up --build` to force it to check if the image needs rebuilding.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
+FROM python:3.7.9-slim-buster as base
```
Comment (hugh-emerson) [inline Dockerfile:1]: You need to update your README with instructions for running the app through Docker.

File: pyproject.toml [hunk 5]
```diff
@@ -8,20 +8,9 @@ authors = []
 python = "^3.7"
 flask = "^1.1.2"
 python-dotenv = "^0.14.0"
```
Comment (hugh-emerson) [inline pyproject.toml:10]: You are missing required packages.

You should have requests, gunicorn, pytest, selenium, pytest-mock.

Ideally the ones just for testing should be in the dev-dependencies section.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+ENV PATH /usr/local/bin:$PATH
+ENV VIRTUAL_ENV "/venv"
+RUN python -m venv $VIRTUAL_ENV
+ENV PATH "$VIRTUAL_ENV/bin:$PATH"
+WORKDIR /app
+RUN pip install poetry
+EXPOSE 5000
+COPY poetry.lock pyproject.toml /app/
+RUN poetry install --no-root
+COPY . /app
```
Comment (hugh-emerson) [inline Dockerfile:12]: If you move some packages into dev-dependencies in pyproject.toml (see my other comment), then you can run `poetry install --no-dev` to skip them.

So you can have a smaller poetry installation for the `production image` if you move this and have different install commands in production/development stages.

**Review by hugh-emerson — COMMENTED** [review]
Generally what you have looks good, but you need to update the README with instructions to run the app with Docker and also fix pyproject.toml. My other comments are just minor suggestions

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+FROM base as production
+ENTRYPOINT poetry run gunicorn -w 4 -b 0.0.0.0:5000 run:app
```
Comment (hugh-emerson) [inline Dockerfile:15]: This will fail because you haven't installed gunicorn - you need to add it to your pyproject.toml

File: .dockerignore [hunk 8]
```diff
@@ -0,0 +1,5 @@
+.env
```
Comment (hugh-emerson) [inline .dockerignore:1]: You should ignore `venv` - the virtual environment created by Poetry.

**Review by hugh-emerson — COMMENTED** [review]
Almost there, you just need to update your .dockerignore, and pyproject.toml in order for this to work reliably.

File: Dockerfile (hunk 7 above)
Comment (wkaiwalya) [reply Dockerfile:15]: I have added the dependency as suggested

**Review by JackMeadDev — APPROVED** [review]
Thanks for making that latest change, I was able to just run your app following the instructions. It's would still be worthwhile adding the `.venv` folder to your dockerignore, but happy for you to merge this when you're ready

**Pull request merged**